oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/CachedGroupMembershipReader.java

```diff
@@ -0,0 +1,209 @@
+            } finally {
+                if (!cacheSuccess || expirationTime == 0) {
+                    synchronized (updates) {
+                        Long exp = updates.get(authorizablePath);
+                        if (Objects.equals(exp, expirationTime)) {
+                            updates.remove(authorizablePath);
+                        }
+                    }
+                }
+            }
+        } else {
+            // load but do not cache. happens when another thread is updating
+            // the cache and this thread is not allowed to serve stale
+            LOG.debug("Load but do not cache. Another thread is updating the cache and this thread is not allowed to serve stale");
+            loader.accept(authorizableNode, groups);
+        }
+    }
+
+    private void serveGroupsFromCache(String authorizablePath,
+                                      Tree principalCache,
+                                      Set<Principal> groups) {
+        LOG.debug("Reading group membership at {}", authorizablePath);
+
+        String str = TreeUtil.getString(principalCache, REP_GROUP_PRINCIPAL_NAMES);
+        if (Strings.isNullOrEmpty(str)) {
+            return;
+        }
+
+        for (String s : Text.explode(str, ',')) {
+            final String name = Text.unescape(s);
+            groups.add(groupPrincipalFactory.create(name));
+        }
+    }
+
+    private void cacheGroups(@NotNull Tree authorizableNode,
+                             @NotNull Set<Principal> groupPrincipals)
+            throws AccessDeniedException, CommitFailedException {
+        try {
+            root.refresh();
+            Tree cache = authorizableNode.getChild(CacheConstants.REP_CACHE);
+            if (!cache.exists()) {
+                if (groupPrincipals.size() <= MEMBERSHIP_THRESHOLD) {
+                    LOG.debug("Omit cache creation for user without group membership at {}", authorizableNode.getPath());
+                    return;
+                } else {
+                    LOG.debug("Create new group membership cache at {}", authorizableNode.getPath());
+                    cache = TreeUtil.addChild(authorizableNode, CacheConstants.REP_CACHE, CacheConstants.NT_REP_CACHE);
+                }
+            }
+
+            cache.setProperty(CacheConstants.REP_EXPIRATION, LongUtils.calculateExpirationTime(expiration));
+            String value = (groupPrincipals.isEmpty()) ? "" : Joiner.on(",").join(Iterables.transform(groupPrincipals, input -> Text.escape(input.getName())));
+            cache.setProperty(REP_GROUP_PRINCIPAL_NAMES, value);
+
+            root.commit(CacheValidatorProvider.asCommitAttributes());
+            LOG.debug("Cached group membership at {}", authorizableNode.getPath());
+        } finally {
+            root.refresh();
+        }
+    }
+
+    private Map<String, Long> getCacheUpdateMap() {
+        return CACHE_UPDATES.computeIfAbsent(config, cfg -> new LinkedHashMap<>(){
```

Comment:
@mreutegg and @nscendoni , had another close look at the PR as i wanted to provide fixes for my picky comments myself.

and while doing that and touching the code in my IDE i realized that i don't understand the nature of the CACHE_UPDATES map that maps UserConfiguration to a second Map<String, Long> which marks the currently being updated authorizable-paths.

Therefore:
Unless i am really mistaken there exists only 1 UserConfiguration for a given oak repository as the SecurityProvider holds only one user configuration. do we need a map with key UserConfiguration?

Reply:
It ensures the cache is per repository and not shared between multiple repository instances. I guess, a better solution would directly tie the cache to a repository instance and pass it down to the CachedGroupMembershipReader.

Reply:
hmmmm.... i don't like the idea of leaking out implementation details of the user management to the whole repository setup. i would like to be able to potentially replace that without having to care about the rest of oak.
so, i will leave it as it is and add a comment explaining the code such that the next reader will understand the intention..